File: moldudp64/test/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"time"

	"moldudp64"
)

const mcAddr = "224.0.0.0:10000"

var (
	session = [10]byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
)

func main() {
	if len(os.Args) == 1 || os.Args[1] != "no" {
		go runTransmitter()
	}

	go runReceiver(1)
	go runReceiver(2)

	time.Sleep(time.Second * 15)
}

func runTransmitter() {
	tm, err := moldudp64.NewTransmitter(mcAddr, session)
	if err != nil {
		log.Fatal(err)
	}
	tm.SetHeartbeat(time.Second * 2)

	time.Sleep(time.Second * 3)

	for i := 1; i <= 10; i++ {
		time.Sleep(time.Second)
		blocks := make([]moldudp64.MessageBlock, 0, i+i)
		for j := 1; j <= cap(blocks); j++ {
			data := []byte(fmt.Sprintf("%d.%d", i, j))
			blocks = append(blocks, must(moldudp64.NewMessageBlock(data)))
		}
		err := tm.SendMessageBlocks(blocks)
		if err != nil {
			log.Fatalf("error sending packet %d: %v", i, err)
		}
	}

	time.Sleep(time.Second)
	if err := tm.SendEndSession(); err != nil {
		log.Fatal("error sending end session: ", err)
	}

	time.Sleep(time.Second)
}

func runReceiver(num int) {
	rcvr, err := moldudp64.NewReceiver(
		mcAddr,
		nil,
		func(packet moldudp64.DownstreamPacket) {
			log.Printf("%d received packet", num)
			for _, block := range packet.MessageBlocks {
				log.Printf("%d received message: %s", num, block.MessageData)
			}
		},
		session,
	)
	if err != nil {
		log.Fatalf("error creating receiver %d: %v", num, err)
	}
	_, _ = <-rcvr.ClosedChan()
	log.Printf("%d closed: %v", num, rcvr.Err())
}

func must[T any](t T, err error) T {
	if err != nil {
		log.Fatal(err)
	}
	return t
}

File: moldudp64/test/main.rs
#[path = "../packet.rs"]
mod packet;
use packet::*;

const ADDR: &str = "224.0.0.0:10000";

fn main() {
    if std::env::args().skip(1).next().map(|s| s != "no").unwrap_or(true) {
        std::thread::spawn(move || run_transmitter());
    }

    std::thread::spawn(move || run_receiver(1));
    std::thread::spawn(move || run_receiver(2));

    std::thread::sleep(std::time::Duration::from_secs(15));
}

fn run_transmitter() {
    let tm = Transmitter::new(ADDR).unwrap();

    for i in 1..=10 {
        std::thread::sleep(std::time::Duration::from_secs(1));
        let mut blocks = Vec::with_capacity(i + 1);
        for j in 1..=blocks.capacity() {
            blocks.push(MessageBlock::new(format!("{i}.{j}").into()).unwrap());
        }
        tm.send(DownstreamPacket::new(blocks).unwrap()).unwrap();
    }

    std::thread::sleep(std::time::Duration::from_secs(1));
    tm.send(DownstreamPacket::end_session()).unwrap();
    std::thread::sleep(std::time::Duration::from_secs(1));
}

fn run_receiver(num: i32) {
    let packet_handler = Box::new(move |packet: DownstreamPacket| {
        println!("{num} received packet");
        for block in packet.message_blocks {
            println!("{num} received message: {}", String::from_utf8_lossy(&block.message_data));
        }
    });
    let rcvr = match Receiver::connect(ADDR, Vec::new(), packet_handler) {
        Ok(r) => r,
        Err(e) => panic!("error creating receiver {}: {}", num, e),
    };
    std::thread::sleep(std::time::Duration::from_secs(15));
}
